runtime/alloc/collect.c

```diff
@@ -107,47 +126,71 @@ static char* get_next(char* scan_ptr, size_t size) {
   return current_tospace_start + sizeof(memory_block_header);
 }
 
+static char* evacuate(char* scan_ptr, char** alloc_ptr) {
+  block *currBlock = (block *)scan_ptr;
+  const uint64_t hdr = currBlock->h.hdr;
+  uint16_t layoutInt = hdr >> 48;
+  if (layoutInt) {
+    layout *layoutData = getLayoutData(layoutInt);
+    for (unsigned i = 0; i < layoutData->nargs; i++) {
+      layoutitem *argData = layoutData->args + i;
+      void *arg = ((char *)currBlock) + argData->offset;
+      switch(argData->cat) {
+      case 1: // map
+        map_foreach(arg, migrate_once);
+	break;
+      case 2: // list
+        list_foreach(arg, migrate_once); 
+	break;
+      case 3: // set
+        set_foreach(arg, migrate_once);
+	break;
+      case 6:  // stringbuffer
+        migrate_string_buffer(arg);
+        break;
+      case 8:  // block
+        migrate(arg);
+        break;
+      case 4: //int
+      case 5: //float
+      case 7: // bool
+      default: //mint
+        break;
+      }
+    }
+  }
+  return get_next(scan_ptr, get_size(hdr, layoutInt), *alloc_ptr);
+}
+
 void koreCollect(block** root) {
   MEM_LOG("Starting garbage collection\n");
   koreAllocSwap();
+  char* oldspace_start = *old_alloc_ptr();
   migrate(root);
   current_tospace_start = fromspace_ptr();
   current_tospace_end = fromspace_ptr() + BLOCK_SIZE;
   char *scan_ptr = current_tospace_start + sizeof(memory_block_header);
+  MEM_LOG("Evacuating young generation\n");
   while(scan_ptr) {
-    block *currBlock = (block *)scan_ptr;
-    const uint64_t hdr = currBlock->h.hdr;
-    uint16_t layoutInt = hdr >> 48;
-    if (layoutInt) {
-      layout *layoutData = getLayoutData(layoutInt);
-      for (unsigned i = 0; i < layoutData->nargs; i++) {
-        layoutitem *argData = layoutData->args + i;
-        void *arg = ((char *)currBlock) + argData->offset;
-        switch(argData->cat) {
-        case 1: // map
-          map_foreach(arg, migrate_once);
-	  break;
-        case 2: // list
-          list_foreach(arg, migrate_once); 
-	  break;
-        case 3: // set
-          set_foreach(arg, migrate_once);
-	  break;
-        case 6:  // stringbuffer
-          migrate_string_buffer(arg);
-          break;
-        case 8:  // block
-          migrate(arg);
-          break;
-        case 4: //int
-        case 5: //float
-        case 7: // bool
-        default: //mint
-          break;
-        }
-      }
+    scan_ptr = evacuate(scan_ptr, alloc_ptr());
+  }
+  scan_ptr = oldspace_start == 0 ? oldspace_ptr() == 0 ? 0 : oldspace_ptr() + sizeof(memory_block_header) : oldspace_start;
```

Comment:
I don't understand this section, beyond the general idea that we want to set current_tospace_start, current_tospace_end and scan_ptr so the scan loop below will evacuate the promoted objects. What invariants are we trying to set up, and what state might we see in these variables and function returns?

Reply:
Maybe this should be extracted as a function like `koreAllocSwap`?